File: go/pkg/trees/trees_2.go
package trees

import (
	"fmt"
	"strings"
)

// for TreeNode type def see Minimal BST

// QueueNode stores a tree node, its depth, and a pointer to the next node
type QueueNode struct {
	TreeNode *TreeNode
	Depth    int
	Next     *QueueNode
}

// ToString QueueNode => string
func (n *QueueNode) ToString() string {
	return fmt.Sprintf("(%d, %d)", n.TreeNode.Data, n.Depth)
}

// Queue is a FIFO data structure
type Queue struct {
	Head   *QueueNode
	Tail   *QueueNode
	Length int
}

// ToString Queue => string
func (q *Queue) ToString() string {
	ss := []string{}
	node := q.Head

	for node != nil {
		ss = append(ss, node.ToString())
		node = node.Next
	}

	return strings.Join(ss, " -> ")
}

// Add a new node in O(1)
func (q *Queue) Add(treeNode *TreeNode, depth int) {
	queueNode := new(QueueNode)
	queueNode.TreeNode = treeNode
	queueNode.Depth = depth

	if q.Head == nil {
		q.Head = queueNode
	}

	if q.Tail != nil {
		q.Tail.Next = queueNode
	}

	q.Tail = queueNode
	q.Length++
}

// Remove a node and return the tree node and depth in O(1)
func (q *Queue) Remove() (*TreeNode, int) {
	if q.Head == nil {
		panic("queue is empty")
	}

	node := q.Head
	q.Head = q.Head.Next
	q.Length--

	return node.TreeNode, node.Depth
}

// ListDepths builds a slice of queues in O(n)
func ListDepths(treeNode *TreeNode, depth int, q *Queue, depths []*Queue) []*Queue {
	// add an additional queue for each depth encountered
	if depth == len(depths) {
		depths = append(depths, new(Queue))
	}

	// add node to queue on depths slice
	depths[depth].Add(treeNode, depth)

	if treeNode.Left != nil {
		q.Add(treeNode.Left, depth+1)
	}

	if treeNode.Right != nil {
		q.Add(treeNode.Right, depth+1)
	}

	if q.Length == 0 {
		return depths
	}

	treeNode, depth = q.Remove()

	return ListDepths(treeNode, depth, q, depths)
}

File: python/trees/trees_2.py
from trees_1 import Node as TreeNode

# for tree node class definition see Minimal BST


class QueueNode:
    def __init__(self, tree_node=None, depth=None, next=None):
        self.tree_node = tree_node
        self.depth = depth
        self.next = next

    def __repr__(self):
        return str((self.tree_node.data, self.depth))


# implemented for O(1) removal of the first element
class Queue:
    def __init__(self):
        self.tail = None
        self.head = None
        self.length = 0

    def __repr__(self):
        nodes = []
        node = self.head

        while node:
            nodes.append(str(node))
            node = node.next

        return ' -> '.join(nodes)

    def add(self, tree_node, depth):
        node = QueueNode(tree_node, depth)

        if not self.head:
            self.head = node

        if self.tail:
            self.tail.next = node

        self.tail = node
        self.length += 1

    def remove(self):
        if not self.head:
            raise Exception('queue is empty')

        node = self.head
        self.head = self.head.next
        self.length -= 1

        return node.tree_node, node.depth


# O(n)
# node_depth is a tuple containing the binary tree node and the depth it was found
# queue is used to implement BFS storing (node, depth) tuples for FIFO access
# depths is a list of queues containing all node data at depth d
# d is both the index of the depths list and the depth in the tree
def list_depths(tree_node, depth, queue=Queue(), depths=[]):
    if depth == len(depths):
        depths.append(Queue())

    depths[depth].add(tree_node, depth)

    if tree_node.left:
        queue.add(tree_node.left, depth + 1)
    if tree_node.right:
        queue.add(tree_node.right, depth + 1)

    if queue.length == 0:
        return depths

    tree_node, depth = queue.remove()
    return list_depths(tree_node, depth, queue, depths)
